### DISCUSSION TITLE
New module `shift_in` and `shift_out` for using shift registers

### DISCUSSION THREAD

**Discussion by Peter-van-Tol** [Peter-van-Tol/LiteX-CNC#52] [Ideas]
When using the 5A-75B, one can get at the point that 51 pins are not enough. The Arduino platform offers two libraries for shifting data using buffers:
- 74HCT595 for output;
- 74HC165 for input;

Using these chip, one can use for example a pendant with loads of functions, connecting the MPG directly to the pins and connecting the buttons to these shift registers.

Design spec:
- each module can handle up to 32 output (`shift_out`) or 32 inputs (`shift_in`);
- pins required are `DATA`, `CLK` and `LATCH`;
- input and output cannot be mixed on one channel;
- support for multiple channels, no overlap in pins. 

Should this prove useful?

**Comment by OJthe123** [comment]
So, this would bei like a port Expander?

**Reply by Peter-van-Tol** [reply]
Exactly!

**Comment by Peter-van-Tol** [comment]
Design challenge is the `shift_in` function. 

The normal routine for LiteX-CNC is:
- read state from FPGA;
- based on the state, update the motion planner and other LinuxCNC components;
- write the new state to the FPGA.

For `shift_out` it is easy. The data is received and directly clocked out to the 74HC595 registers, so the response time of the GPIO will be in order magnitude <1 ms.

For `shift in` we have to determine when to read the registers. Proposed solution is now to send a flag while writing the data to the FPGA, which starts the data acquisition for the next read in the cycle. This means that the data can be 'as old as' 1 ms.

**Comment by OJthe123** [comment]
Since all real-time data can be connected directly to the fpga, a 1ms delay for non timing critical events, like a pendant, etc would be no problem I think.

**Comment by Peter-van-Tol** [comment]
The writing of these components are finally under way, see #86 .

Currently the 74HC595 (`shift out`) register is supported, but yet untested and undocumented. After testing I will start with adding the 74HC165 register (`shift_in`). The chosen architecture means that inputs and outputs cannot be mixed. This means you must have separate data-lines for each of the modules. 

An practical example, a pendant requiring E-stop, input, output and a MPG, would require:
- 1 wire for the E-stop (ideally this would be a two wire). This E-stop is wired directly into the real-time components;
- 1 wire for enabling the pendant.
- 4 wires for input (`shift_in`), which are CLOCK, DATA, LATCH and CLEAR.
- 4 wires for output (`shift_out`), which are CLOCK, DATA, LATCH and CLEAR.
- 4 wires for the MPG, which are A-/A+ and B-/B+, assuming ,you want to run the MPG with differential signal.
- 2 wires for power.

With this you can have endless buttons and LED's on your MPG (ok, it is limited to 256), but the number of wires is still larger then using a serial protocol (which is 4 wires).

Version 1.3 of LitexCNC is upcoming, hopefully I can include this feature in this version.
